### PR TITLE
Add an initial MOPAC parser based on RMG-Py contribution

### PR DESCRIPTION
**Pull request by ghutchis**
I am starting an initial MOPAC output parser. I will add support for as much parsed data as possible.

I'm curious at the start about coding standards, test files, etc. for contributed parsers. (Is there a guide somewhere .. "if you're contributing a parser, here's what you need?")

### CODE REVIEW

**Comment by berquist** [comment]
So @ben-albrecht started this, but I don't know how far he got; his initial work is directly in his master branch [here](https://github.com/ben-albrecht/cclib). This may or may not be of use to you.

> Is there a guide somewhere .. "if you're contributing a parser, here's what you need?"

Unfortunately, no. It mostly involves copying an existing one as a starting point for `mopacparser.py` and making an equivalent set of test inputs and outputs as best as possible which would reside in `data/MOPAC/basicMOPAC2012`. The list of all regular test inputs and classes is in [this file](https://github.com/cclib/cclib/blob/master/test/testdata). The test classes themselves are in [this directory](https://github.com/cclib/cclib/tree/master/test/data). Not every test is applicable to every program, which is why there isn't uniform coverage. To run the test suite, [here](https://gist.github.com/berquist/b0c2c5004f1ac25748ce) is a script that can be placed in the top-level directory and executed. To only run the basic and regression tests for a single parser, the `test_data` and `run_regressions` lines accept the name of one or more parsers as arguments. It's fine to not cover everything MOPAC is capable of right away. 

The only rules that I know of are that we don't break trunk, the more modern code is usually PEP8-compliant (though the line lengths are flexible), and everything is both Python 2 and 3 compatible.

I know that MOPAC creates multiple output files, but I can't remember if there's any non-redundant information in the secondary one(s), like the *.arc files. If you're interested in parsing those as well, the Molpro parser shows how to do it.

**Comment by ghutchis** [comment]
I will try to merge the parser from @ben-albrecht and the older one. @ATenderholt had suggested I start from the mopac branch (which is older). The West group, that wrote the older parser said they had some changes too.

Parsing the .arc files would be useful, since the precision is sometimes better. Stewart's declaration is that the .out is for humans and .arc is for programs.

Thanks for the other pointers - perhaps it's worth putting some of these in the docs or the [wiki](https://github.com/cclib/cclib/wiki).

**Comment by ben-albrecht** [comment]
I didn't make it very far past parsing charge, multiplicity, geometry, and atom elements. You can grab that from my attempt at [mopacparser.py](https://github.com/ben-albrecht/cclib/blob/master/cclib/parser/mopacparser.py) if it is helpful. There are also some tests set up.

Ignore the huge comment blocks - I began with @berquist's Q-Chem parser as a guide to work off of, and left most of the commented code in while working on it.

H2P

**Comment by langner** [comment]
Yeah, there isn't really anything documenting how to add a new parser. It has happened infrequently enough to not really be an issue.

Creating the parser is usually the easy part - just make sure it inherits the Logfile class and implements some of the methods needed. Getting good coverage of attributes and having working unit tests is more work.

**Comment by langner** [comment]
Could you also add at the SP unit test here?

**Comment by ghutchis** [comment]
Yes, I have a work-in-progress on my laptop with a more complete parser, unit test, and fix for the break indicated by Travis. I'll probably post later today.

**Comment by langner** [comment]
Great! I marked this for the next release.

**Comment by ghutchis** [comment]
Still need to add some MOPAC tests

File: src/cclib/parser/utils.py [hunk 1]
```diff
@@ -111,7 +112,8 @@ def __init__(self):
             'Tl', 'Pb', 'Bi', 'Po', 'At', 'Rn',
             'Fr', 'Ra',
             'Ac', 'Th', 'Pa', 'U', 'Np', 'Pu', 'Am', 'Cm', 'Bk', 'Cf', 'Es', 'Fm', 'Md', 'No',
-            'Lr', 'Rf', 'Db', 'Sg', 'Bh', 'Hs', 'Mt', 'Ds', 'Rg', 'Uub']
+            'Lr', 'Rf', 'Db', 'Sg', 'Bh', 'Hs', 'Mt', 'Ds', 'Rg', 'Cn',
+            'Uut', 'Fl', 'Uup', 'Lv', 'Uus', 'Uuo']
```
Comment (berquist) [inline src/cclib/parser/utils.py]: We have new element names now!

**Comment by berquist** [comment]
One thing that I know Ben created was inputs and outputs for the "standard" test files we use with divinylbenzene, which helps us with test reuse. There will be specific cases for MOPAC due to it being semiempirical but that can't be helped. Can you create and reuse those inputs? The test suite can be modified as we go.

`data/MOPAC/benzene-stop.res` is a binary file that shouldn't be included.

**Comment by ghutchis** [comment]
I'll be happy to look around for the DVB file, sure. Just let me know what the "standard" files are and I'll run them - shouldn't take more than a few minutes. (The specific issue with MOPAC is more that it uses STOs.)

**Comment by langner** [comment]
I think the minimum is to have single point and geometry optimization for DVB. You can take the input geometries from the unit tests of any other programs. For example:
https://github.com/cclib/cclib/blob/master/data/Gaussian/basicGaussian09/dvb_gopt.gjf (Z-matrix)
https://github.com/cclib/cclib/blob/master/data/GAMESS/basicGAMESS-US2012/dvb_gopt_a.inp (XYZ)

**Comment by langner** [comment]
@ghutchis why are STO basis sets an issue?

**Comment by ghutchis** [comment]
Does cclib have support for storing STO basis set information (or translating to GTO)?

**Comment by langner** [comment]
Oh, I just understood that by STO you mean real Slater type orbitals, not GTO fitted to STOs (as in STO-3G, etc.).

I suppose we currently _assume_ the basis set is Gaussian. But we do parse ADF output, which uses STOs and "fragment orbitals", so the precedent is there. Although cclib does not parse the STO parameters from ADF output currently.

**Comment by ghutchis** [comment]
OK, I added the DVB test files and will work later to add some tests. I think for now, I'll punt on STO parsing, but try to get most of the other data.

**Comment by langner** [comment]
@ghutchis Do you think this is ready for v1.5? Would you manage to resolve the conflicts before Sunday?

**Comment by ghutchis** [comment]
I'll resolve the conflicts. I think there's a useful start of a parser. It's obviously not complete, but should be useful.

**Comment by berquist** [comment]
Yes, we definitely want this; some "high-profile" users were talking about it online just a few days ago.

**Comment by langner** [comment]
> Yes, we definitely want this; some "high-profile" users were talking about it online just a few days ago.

link? :)

**Comment by langner** [comment]
> I'll resolve the conflicts. I think there's a useful start of a parser. It's obviously not complete, but should be useful.

Great, thanks. It'll be great to have the new parser in a new version.

**Comment by berquist** [comment]
See [this](https://twitter.com/janhjensen/status/765876193820311553) thread.

**Comment by langner** [comment]
> Thanks for the other pointers - perhaps it's worth putting some of these in the docs or the wiki.

We store the docs in a separate directory (https://github.com/cclib/cclib.github.io), but I agree, would be good to update them with this information. Put some things in cclib/cclib.github.io#20 for updating the docs.

**Comment by ghutchis** [comment]
I'm beyond my level here:
```
======================================================================
ERROR: test_data_stdin (testlogfileparser.FileWrapperTest)
Check that the same attributes are parsed when a file is piped through standard input.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "parser/testlogfileparser.py", line 81, in test_data_stdin
    expected_attributes = get_attributes(cclib.io.ccopen(path).parse())
AttributeError: 'module' object has no attribute 'io'

======================================================================
FAIL: test_stdin_seek (testlogfileparser.FileWrapperTest)
We shouldn't be able to seek anywhere in standard input.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "parser/testlogfileparser.py", line 69, in test_stdin_seek
    self.assertRaises(IOError, wrapper.seek, 0, 0)
AssertionError: IOError not raised

----------------------------------------------------------------------
Ran 17 tests in 0.713s

FAILED (failures=1, errors=1)
````

**Comment by ghutchis** [comment]
I tried to fix conflicts and the merge looked OK, but at this point, the tests look seriously confusing. I'm happy to revert that merge and let someone else do it.

**Comment by langner** [comment]
I'll take a look in a bit.

**Comment by langner** [comment]
Please see ghutchis/cclib#1

**Comment by ghutchis** [comment]
I'm using Python 2.7.11 :: Anaconda 2.3.0 (x86_64)
but the check looks OK and it works for me. Free free to merge if it looks good to you.

**Comment by langner** [comment]
Yup, looks good.

**Pull request merged**